File: experiments/adrianco/experiment-27-sampling-ff/bookshop/runs/agent=hermes-local_language=go_prompt=none_stack=s8/rep1/handlers/books.go
package handlers

import (
	"book-api/db"
	"encoding/json"
	"net/http"
	"strconv"
	"strings"

	"github.com/gorilla/mux"
)

// BooksHandler handles HTTP requests for book operations.
type BooksHandler struct {
	Db *db.Database
}

// NewBooksHandler creates a new handler instance.
func NewBooksHandler(database *db.Database) *BooksHandler {
	return &BooksHandler{Db: database}
}

// HealthCheck handles the /health endpoint.
func (h *BooksHandler) HealthCheck(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(map[string]interface{}{
		"status": "healthy",
	})
}

// CreateBook handles POST /books.
func (h *BooksHandler) CreateBook(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Title  string `json:"title"`
		Author string `json:"author"`
		Year   int    `json:"year"`
		ISBN   string `json:"isbn"`
	}

	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, `{"error":"invalid request body"}`, http.StatusBadRequest)
		return
	}

	if req.Title == "" {
		http.Error(w, `{"error":"title is required"}`, http.StatusBadRequest)
		return
	}
	if req.Author == "" {
		http.Error(w, `{"error":"author is required"}`, http.StatusBadRequest)
		return
	}

	book, err := h.Db.CreateBook(req.Title, req.Author, req.Year, req.ISBN)
	if err != nil {
		http.Error(w, `{"error":"failed to create book"}`, http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusCreated)
	json.NewEncoder(w).Encode(book)
}

// GetBooks handles GET /books and GET /books?author=...
func (h *BooksHandler) GetBooks(w http.ResponseWriter, r *http.Request) {
	authorFilter := r.URL.Query().Get("author")

	books, err := h.Db.GetAllBooks(authorFilter)
	if err != nil {
		http.Error(w, `{"error":"failed to retrieve books"}`, http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(books)
}

// GetBook handles GET /books/{id}.
func (h *BooksHandler) GetBook(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	idStr := vars["id"]

	id, err := strconv.Atoi(idStr)
	if err != nil {
		http.Error(w, `{"error":"invalid book ID"}`, http.StatusBadRequest)
		return
	}

	book, err := h.Db.GetBookByID(id)
	if err != nil {
		http.Error(w, `{"error":"book not found"}`, http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(book)
}

// UpdateBook handles PUT /books/{id}.
func (h *BooksHandler) UpdateBook(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	idStr := vars["id"]

	id, err := strconv.Atoi(idStr)
	if err != nil {
		http.Error(w, `{"error":"invalid book ID"}`, http.StatusBadRequest)
		return
	}

	var req map[string]interface{}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, `{"error":"invalid request body"}`, http.StatusBadRequest)
		return
	}

	if title, ok := req["title"].(string); ok {
		if strings.TrimSpace(title) == "" {
			http.Error(w, `{"error":"title cannot be empty"}`, http.StatusBadRequest)
			return
		}
	}
	if author, ok := req["author"].(string); ok {
		if strings.TrimSpace(author) == "" {
			http.Error(w, `{"error":"author cannot be empty"}`, http.StatusBadRequest)
			return
		}
	}

	var title *string
	if v, ok := req["title"].(string); ok {
		title = &v
	}
	var author *string
	if v, ok := req["author"].(string); ok {
		author = &v
	}
	var year *int
	if v, ok := req["year"].(float64); ok {
		vInt := int(v)
		year = &vInt
	}
	var isbn *string
	if v, ok := req["isbn"].(string); ok {
		isbn = &v
	}

	book, err := h.Db.UpdateBook(id, title, author, year, isbn)
	if err != nil {
		http.Error(w, `{"error":"book not found"}`, http.StatusNotFound)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(book)
}

// DeleteBook handles DELETE /books/{id}.
func (h *BooksHandler) DeleteBook(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	idStr := vars["id"]

	id, err := strconv.Atoi(idStr)
	if err != nil {
		http.Error(w, `{"error":"invalid book ID"}`, http.StatusBadRequest)
		return
	}

	if err := h.Db.DeleteBook(id); err != nil {
		http.Error(w, `{"error":"book not found"}`, http.StatusNotFound)
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

File: experiments/adrianco/experiment-38-alllang-80b-fullctx/bookshop/runs/agent=hermes-local_language=typescript_model=mlxlocal/mlx-community--Qwen3-Coder-Next-4bit_prompt=neutral_stack=m80/rep3/src/controllers/books.ts
import { Request, Response } from 'express';
import { BookInput } from '../types';
import { getAllBooks, getBookById, createBook, updateBook, deleteBook } from '../models/book';

export async function getAllBooksHandler(req: Request, res: Response): Promise<void> {
  try {
    const author = req.query.author as string | undefined;
    const books = await getAllBooks(author);
    res.json(books);
  } catch (error) {
    res.status(500).json({ error: 'Failed to retrieve books' });
  }
}

export async function getBookHandler(req: Request, res: Response): Promise<void> {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) {
      res.status(400).json({ error: 'Invalid book ID' });
      return;
    }
    
    const book = await getBookById(id);
    if (!book) {
      res.status(404).json({ error: 'Book not found' });
      return;
    }
    
    res.json(book);
  } catch (error) {
    res.status(500).json({ error: 'Failed to retrieve book' });
  }
}

export async function createBookHandler(req: Request, res: Response): Promise<void> {
  try {
    const book: BookInput = req.body;
    const newBook = await createBook(book);
    res.status(201).json(newBook);
  } catch (error) {
    res.status(500).json({ error: 'Failed to create book' });
  }
}

export async function updateBookHandler(req: Request, res: Response): Promise<void> {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) {
      res.status(400).json({ error: 'Invalid book ID' });
      return;
    }
    
    const book: BookInput = req.body;
    const updatedBook = await updateBook(id, book);
    
    if (!updatedBook) {
      res.status(404).json({ error: 'Book not found' });
      return;
    }
    
    res.json(updatedBook);
  } catch (error) {
    res.status(500).json({ error: 'Failed to update book' });
  }
}

export async function deleteBookHandler(req: Request, res: Response): Promise<void> {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) {
      res.status(400).json({ error: 'Invalid book ID' });
      return;
    }
    
    const deleted = await deleteBook(id);
    if (!deleted) {
      res.status(404).json({ error: 'Book not found' });
      return;
    }
    
    res.status(204).send();
  } catch (error) {
    res.status(500).json({ error: 'Failed to delete book' });
  }
}
